File: couple-carrier-risk/r/tests/testthat/test-orchestrator.R
make_fake_client <- function() {
  state <- new.env(parent = emptyenv())
  state$next_id <- 0L
  state$created_pedigrees <- character(0)
  state$created_individuals <- list()
  state$memberships <- list()
  state$imported <- list()
  state$deleted_pedigrees <- character(0)
  state$deleted_individuals <- character(0)
  state$recorded_ancestries <- list()
  state$ancestry_lookup <- list()
  state$get_population_risks <- NULL

  fresh_id <- function(prefix) {
    state$next_id <- state$next_id + 1L
    sprintf("%s-%d", prefix, state$next_id)
  }

  client <- list(
    create_pedigree = function(display_name) {
      state$created_pedigrees <- c(state$created_pedigrees, display_name)
      fresh_id("ped")
    },
    delete_pedigree = function(pedigree_id) {
      state$deleted_pedigrees <- c(state$deleted_pedigrees, pedigree_id)
    },
    add_individual_to_pedigree = function(pedigree_id, individual_id) {
      state$memberships[[length(state$memberships) + 1L]] <- c(pedigree_id, individual_id)
    },
    create_individual = function(display_name, biological_sex) {
      identifier <- fresh_id("ind")
      state$created_individuals[[length(state$created_individuals) + 1L]] <- list(
        id = identifier,
        display_name = display_name,
        biological_sex = biological_sex
      )
      identifier
    },
    delete_individual = function(individual_id) {
      state$deleted_individuals <- c(state$deleted_individuals, individual_id)
    },
    import_23andme_raw = function(pedigree_id, individual_id, tsv) {
      state$imported[[length(state$imported) + 1L]] <- list(id = individual_id)
    },
    find_ancestry_id_by_population_key = function(population_key) {
      state$ancestry_lookup[[population_key]]
    },
    add_ancestry_to_individual = function(individual_id, ancestry_id) {
      state$recorded_ancestries[[length(state$recorded_ancestries) + 1L]] <-
        c(individual_id, ancestry_id)
    },
    get_population_risks = function(individual_id) {
      if (!is.null(state$get_population_risks)) {
        return(state$get_population_risks(individual_id))
      }
      list(
        individual_id = individual_id,
        risks = list(list(
          disease_id = "d1",
          disease_name = "Sickle cell anaemia",
          inheritance_pattern = "autosomal_recessive",
          carrier_frequency = 0.05,
          couple_offspring_risk = 0.000625
        ))
      )
    }
  )
  list(client = client, state = state)
}

test_config <- function(ancestry_a = "auto", ancestry_b = "auto", cleanup = TRUE) {
  list(
    base_url = "https://evagene.example",
    api_key = "evg_test",
    partner_a_file = fixture_path("partner-a-23andme.txt"),
    partner_b_file = fixture_path("partner-b-23andme.txt"),
    ancestry_a = ancestry_a,
    ancestry_b = ancestry_b,
    output = "table",
    cleanup = cleanup
  )
}

capture_output <- function(fn) {
  output <- character(0)
  sink_conn <- textConnection("output", "w", local = TRUE)
  on.exit(close(sink_conn), add = TRUE)
  fn(sink_conn)
  paste(output, collapse = "\n")
}

test_that("end-to-end happy path creates scratch and renders the table", {
  fake <- make_fake_client()

  output <- capture_output(function(sink) {
    run_couple_screening(test_config(), fake$client, sink)
  })

  expect_true(grepl("Sickle cell anaemia", output, fixed = TRUE))
  expect_equal(fake$state$created_pedigrees, SCRATCH_PEDIGREE_NAME)
  expect_equal(
    vapply(fake$state$created_individuals, function(ind) ind$display_name, character(1L)),
    c("Partner A", "Partner B")
  )
  expect_length(fake$state$imported, 2L)
})

test_that("cleanup deletes pedigree and individuals on success", {
  fake <- make_fake_client()

  capture_output(function(sink) {
    run_couple_screening(test_config(cleanup = TRUE), fake$client, sink)
  })

  expect_length(fake$state$deleted_individuals, 2L)
  expect_length(fake$state$deleted_pedigrees, 1L)
})

test_that("--no-cleanup leaves scratch in place", {
  fake <- make_fake_client()

  capture_output(function(sink) {
    run_couple_screening(test_config(cleanup = FALSE), fake$client, sink)
  })

  expect_length(fake$state$deleted_pedigrees, 0L)
  expect_length(fake$state$deleted_individuals, 0L)
})

test_that("cleanup runs even when risk fetch fails", {
  fake <- make_fake_client()
  fake$state$get_population_risks <- function(individual_id) {
    stop("network broke mid-run")
  }

  expect_error(
    capture_output(function(sink) {
      run_couple_screening(test_config(cleanup = TRUE), fake$client, sink)
    }),
    "network broke"
  )

  expect_gte(length(fake$state$deleted_pedigrees), 1L)
})

test_that("explicit ancestry is looked up and attached", {
  fake <- make_fake_client()
  fake$state$ancestry_lookup <- list(mediterranean = "anc-uuid")

  capture_output(function(sink) {
    run_couple_screening(
      test_config(ancestry_a = "mediterranean", ancestry_b = "mediterranean"),
      fake$client, sink
    )
  })

  expect_equal(
    vapply(fake$state$recorded_ancestries, function(entry) entry[[2L]], character(1L)),
    c("anc-uuid", "anc-uuid")
  )
})

test_that("unknown ancestry key raises ancestry_not_found_error", {
  fake <- make_fake_client()
  fake$state$ancestry_lookup <- list()

  expect_error(
    capture_output(function(sink) {
      run_couple_screening(
        test_config(ancestry_a = "klingon"), fake$client, sink
      )
    }),
    class = "ancestry_not_found_error"
  )
})

File: couple-carrier-risk/python/tests/test_orchestrator.py
import io
from typing import Any

import pytest

from couple_carrier_risk.config import Config
from couple_carrier_risk.evagene_client import EvageneClient
from couple_carrier_risk.orchestrator import (
    SCRATCH_PEDIGREE_NAME,
    AncestryNotFoundError,
    run_couple_screening,
)
from tests.fixtures_loader import fixture_path


def _config(ancestry_a: str = "auto", ancestry_b: str = "auto", cleanup: bool = True) -> Config:
    return Config(
        base_url="https://evagene.example",
        api_key="evg_test",
        partner_a_file=str(fixture_path("partner-a-23andme.txt")),
        partner_b_file=str(fixture_path("partner-b-23andme.txt")),
        ancestry_a=ancestry_a,
        ancestry_b=ancestry_b,
        output_format="table",
        cleanup=cleanup,
    )


class _FakeClient:
    """Captures every call the orchestrator makes and returns canned responses."""

    def __init__(self) -> None:
        self.created_pedigrees: list[str] = []
        self.created_individuals: list[tuple[str, str]] = []  # (display_name, sex)
        self.pedigree_memberships: list[tuple[str, str]] = []
        self.imported_tsvs: list[tuple[str, str]] = []
        self.deleted_pedigrees: list[str] = []
        self.deleted_individuals: list[str] = []
        self.recorded_ancestries: list[tuple[str, str]] = []
        self.ancestry_lookup: dict[str, str | None] = {}
        self._next_id = 0

    def _fresh_id(self, prefix: str) -> str:
        self._next_id += 1
        return f"{prefix}-{self._next_id}"

    def create_pedigree(self, display_name: str) -> str:
        identifier = self._fresh_id("ped")
        self.created_pedigrees.append(display_name)
        return identifier

    def delete_pedigree(self, pedigree_id: str) -> None:
        self.deleted_pedigrees.append(pedigree_id)

    def add_individual_to_pedigree(self, pedigree_id: str, individual_id: str) -> None:
        self.pedigree_memberships.append((pedigree_id, individual_id))

    def create_individual(self, *, display_name: str, biological_sex: Any) -> Any:
        identifier = self._fresh_id("ind")
        self.created_individuals.append((display_name, biological_sex.value))
        from couple_carrier_risk.evagene_client import Individual
        return Individual(id=identifier, display_name=display_name)

    def delete_individual(self, individual_id: str) -> None:
        self.deleted_individuals.append(individual_id)

    def import_23andme_raw(self, *, pedigree_id: str, individual_id: str, tsv: str) -> None:
        self.imported_tsvs.append((individual_id, tsv[:20]))

    def find_ancestry_id_by_population_key(self, population_key: str) -> str | None:
        return self.ancestry_lookup.get(population_key)

    def add_ancestry_to_individual(
        self, *, individual_id: str, ancestry_id: str, proportion: float = 1.0,
    ) -> None:
        self.recorded_ancestries.append((individual_id, ancestry_id))

    def get_population_risks(self, individual_id: str) -> dict[str, Any]:
        return {
            "individual_id": individual_id,
            "risks": [
                {
                    "disease_id": "d1",
                    "disease_name": "Sickle cell anaemia",
                    "inheritance_pattern": "autosomal_recessive",
                    "carrier_frequency": 0.05,
                    "couple_offspring_risk": 0.000625,
                },
            ],
        }


def test_end_to_end_happy_path_creates_scratch_and_renders_table() -> None:
    client = _FakeClient()
    sink = io.StringIO()

    run_couple_screening(_config(), client, sink)  # type: ignore[arg-type]

    output = sink.getvalue()
    assert "Sickle cell anaemia" in output
    assert client.created_pedigrees == [SCRATCH_PEDIGREE_NAME]
    assert [entry[0] for entry in client.created_individuals] == ["Partner A", "Partner B"]
    assert len(client.imported_tsvs) == 2


def test_cleanup_deletes_pedigree_and_individuals_on_success() -> None:
    client = _FakeClient()

    run_couple_screening(_config(cleanup=True), client, io.StringIO())  # type: ignore[arg-type]

    assert len(client.deleted_individuals) == 2
    assert len(client.deleted_pedigrees) == 1


def test_no_cleanup_flag_leaves_scratch_in_place() -> None:
    client = _FakeClient()

    run_couple_screening(_config(cleanup=False), client, io.StringIO())  # type: ignore[arg-type]

    assert client.deleted_pedigrees == []
    assert client.deleted_individuals == []


def test_cleanup_runs_even_when_risk_fetch_fails() -> None:
    class _FlakyClient(_FakeClient):
        def get_population_risks(self, individual_id: str) -> dict[str, Any]:
            raise RuntimeError("network broke mid-run")

    client = _FlakyClient()

    with pytest.raises(RuntimeError, match="network broke"):
        run_couple_screening(_config(cleanup=True), client, io.StringIO())  # type: ignore[arg-type]

    assert client.deleted_pedigrees, "pedigree must still be deleted on failure"


def test_explicit_ancestry_is_looked_up_and_attached() -> None:
    client = _FakeClient()
    client.ancestry_lookup = {"mediterranean": "anc-uuid"}

    run_couple_screening(
        _config(ancestry_a="mediterranean", ancestry_b="mediterranean"),
        client,  # type: ignore[arg-type]
        io.StringIO(),
    )

    assert [entry[1] for entry in client.recorded_ancestries] == ["anc-uuid", "anc-uuid"]


def test_unknown_ancestry_key_raises_ancestry_not_found() -> None:
    client = _FakeClient()
    client.ancestry_lookup = {}

    with pytest.raises(AncestryNotFoundError):
        run_couple_screening(
            _config(ancestry_a="klingon"),
            client,  # type: ignore[arg-type]
            io.StringIO(),
        )


def test_fake_client_implements_evagene_client_surface() -> None:
    """Smoke-check: every method used by the orchestrator exists on EvageneClient."""
    required = {
        "create_pedigree",
        "delete_pedigree",
        "add_individual_to_pedigree",
        "create_individual",
        "delete_individual",
        "import_23andme_raw",
        "find_ancestry_id_by_population_key",
        "add_ancestry_to_individual",
        "get_population_risks",
    }
    for name in required:
        assert hasattr(EvageneClient, name), f"EvageneClient missing {name!r}"
